### PR TITLE
[Stack Monitoring] Adjust parity test to pull out correct ILM entry

### PR DESCRIPTION
**Pull request by cachedout**
Parity tests began to fail recently and it appears that the problem is that an extra ILM policy is being created. I don't think there is a problem with the parity itself and this is just something that needs to be accounted for in the logic which compares documents. I found that previously, we were just popping off the last document but that this no longer works because the documents are ordered as they previously were.

This simply adjusts the logic to pull out the extra ILM policy from the Metricbeat document in the correct position.

I am not sure of the merge-forward policy in this repo, so this may need a follow-up PR to the master branch after this is reviewed.

### CODE REVIEW

**Comment by ycombinator** [comment]
Thanks for the fix @cachedout.

It's curious that this ILM-policy-related failure is only happening on the `7.x` branch. Typically `7.x` tends to be pretty close to `master`. 

Let me take a closer look at where this extra policy is coming from and why it is (apparently) only being created in `7.x` but not `master`.

**Comment by cachedout** [comment]
Sounds good. I didn't get a chance to look at this deeply but it's worth noting that the failures happened within a pretty small testing window. You might be able to eyeball the commit history for possible culprits. 

First Failure: Aug 29, 2019 8:13 AM

Last Success: Aug 28, 2019 11:56 AM

**Comment by ycombinator** [comment]
Just as a note of interest, the first failure that happened on Aug 29 was using `ES_BUILD_URL=snapshots.elastic.co/7.4.0-548b3eaa`.  So this is related to some change that was introduced in 7.4 itself, not just 7.5 and later. 

Still investigating what the actual change might be...

File: playbooks/monitoring/elasticsearch/docs_compare.py [hunk 1]
```diff
@@ -94,7 +94,7 @@ def handle_special_case_cluster_stats(internal_doc, metricbeat_doc):
     # usage stats associated with the Metricbeat-created ILM policy.
     ilm = metricbeat_doc["stack_stats"]["xpack"]["ilm"]
 
-    ilm["policy_stats"].pop()
```
Comment (ycombinator) [inline playbooks/monitoring/elasticsearch/docs_compare.py]: I've been looking into this more deeply and I can't figure out why `pop(1)` works! 

Working off the first failing build's release snapshot, here's what the `metricbeat_doc["stack_stats"]["xpack"]["ilm"]["policy_stats"]` list looks like, before it is manipulated in the `handle_special_case_cluster_stats` function:

```
[{u'phases': {u'hot': {u'actions': [u'rollover'], u'min_age': 0}}, u'indices_managed': 1}, {u'phases': {u'delete': {u'actions': [u'delete'], u'min_age': 604800000.0}}, u'indices_managed': 0}, {u'phases': {u'hot': {u'actions': [u'rollover'], u'min_age': 0}, u'delete': {u'actions': [u'delete'], u'min_age': 7776000000.0}}, u'indices_managed': 0}]
```

After `pop(1)` is applied to this list, it looks like this:

```
[{u'phases': {u'hot': {u'actions': [u'rollover'], u'min_age': 0}}, u'indices_managed': 1}, {u'phases': {u'hot': {u'actions': [u'rollover'], u'min_age': 0}, u'delete': {u'actions': [u'delete'], u'min_age': 7776000000.0}}, u'indices_managed': 0}]
```


And here's what the `internal_doc["stack_stats"]["xpack"]["ilm"]["policy_stats"]` list looks like:

```
[{u'phases': {u'delete': {u'actions': [u'delete'], u'min_age': 604800000}}, u'indices_managed': 0}, {u'phases': {u'hot': {u'actions': [u'rollover'], u'min_age': 0}, u'delete': {u'actions': [u'delete'], u'min_age': 7776000000}}, u'indices_managed': 0}]
```

If you compare the second and third lists above, they shouldn't pass the parity check. Specifically, if you look at the first element in each list, the second list (which is part of `metricbeat_doc`) has `phases.hot` but the third one (which is part of `internal_doc`) has `phases.delete`.

So I would not expect these to lists to have parity - yet the script says they do!

So I did some manual testing by performing a `diff` on just the relevant parts of the lists:

```
$ python
Python 2.7.15 (default, Jun 17 2018, 12:46:58)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from jsondiff import diff
>>> i = [ { 'phases': { 'delete': { 'actions': [ 'delete' ] } } } ]
>>> m = [ { 'phases': { 'hot': { 'actions': [ 'rollover' ] } } } ]
>>> diff(i,m, syntax='explicit', marshal=True)
{0: {'$update': {'phases': {'hot': {'actions': ['rollover']}}}}}
```

I would've expected the diff to produce something along the lines of:

```
{0: {'phases': {'$delete': ['delete'], '$insert': ['hot']}}}
```

This makes me _really_ nervous about the [`diff` utility](https://github.com/xlwings/jsondiff) we're using all over these parity tests.

WDYT?

Comment (cachedout) [reply playbooks/monitoring/elasticsearch/docs_compare.py]: What follows is my initial thoughts on the final part of your findings, where you say that you would have expected the diff to report: `{0: {'phases': {'$delete': ['delete'], '$insert': ['hot']}}}`

Maybe I misunderstand the nature of your concern, but this actually seems right to me. Let me break into some hypothetical examples to try to explain how I see it.
```
>>> a
{'a': {'z': 'c'}}
>>> b
{'a': {'b': 'd'}}
>>> diff(a,b, syntax='explicit', marshal=True)
{'$update': {'a': {'b': 'd'}}}
```
The above seems to me like it's reporting diffs at the highest level of divergence in the structure and it sounds to me like you're expecting it to be reporting differences deeper in the structure instead. (I'm assuming that so please correct me if I'm wrong!) 

Here's a further example of how I see this working:

```
>>> a = {'a': {'y': 'c'}}
>>> b = {'a': {'z': 'd'}}
>>> diff(a,b, syntax='explicit', marshal=True)
{'$update': {'a': {'z': 'd'}}}

>>> a = {'a': {'b': 'c'}}
>>> b = {'a': {'b': 'd'}}
>>> diff(a,b, syntax='explicit', marshal=True)
{'$update': {'a': {'$update': {'b': 'd'}}}}
```

In the above first example, we get what I'd expect, though I do think the _way_ it reports this is a bit weird. I read the first return to be in plain English: "There is a difference in the value of 'a', and that difference is not the existence or absence of a value for 'a' (i.e., an insertion or deletion) but instead a difference that we'll call an "update")".

In the second example, where we've partially corrected the first example so that we now have a common key ('b'), it again reports what I would expect, which is that now we have a parity issue deeper in the structure.

I have a meeting now but after I'll come back and look more at the first part of your findings. Again, if I've totally mischaracterized your concern please let me know. Thanks!

Comment (ycombinator) [reply playbooks/monitoring/elasticsearch/docs_compare.py]: I think your first example characterizes it well. The fact that the utility reports the change as an "update" is what's causing the problem in the parity test code, I think. 

Ultimately I think what'd the parity tests should know about is that "there was a `a.z` key in `a` that doesn't exist in `b`". If `a` represents the internally-collected document and `b` represents the metricbeat-collected document, I think knowing about missing properties like this would be significant.

Given what we've discovered about how the `diff` utility reports things, I think the parity tests are currently missing out on detecting such "missing property" differences. WDYT?

Comment (cachedout) [reply playbooks/monitoring/elasticsearch/docs_compare.py]: That seems correct to me. However, unless I'm missing something that still doesn't explain the mystery of why the parity test started failing in the first place, which is still really curious.

Comment (ycombinator) [reply playbooks/monitoring/elasticsearch/docs_compare.py]: > That seems correct to me. 

Cool. In that case, I'm okay with accepting this PR as-is (will approve it in a second) and filing a separate issue to take a closer look at how differences are calculated in parity tests.

> However, unless I'm missing something that still doesn't explain the mystery of why the parity test started failing in the first place, which is still really curious.

Sorry, I forgot to include this bit earlier. It looks like [SLM](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/getting-started-snapshot-lifecycle-management.html) introduced a new ILM policy, which is what messed up the ordering in the policy usage stats.

Comment (ycombinator) [reply playbooks/monitoring/elasticsearch/docs_compare.py]: Here's the issue mentioned in the previous comment: https://github.com/elastic/elastic-stack-testing/issues/315.

**Comment by ycombinator** [comment]
> I am not sure of the merge-forward policy in this repo, so this may need a follow-up PR to the master branch after this is reviewed.

++ I expect parity tests in `master` to fail with the same error, as soon as we have a good release snapshot build on `master`.

**Comment by ycombinator** [comment]
@cachedout Now that `7.x` points to `7.5`, you will need to backport this PR to the `7.4` branch as well.
